### ISSUE TITLE
Container with margin containing expanded DefaultTabController throws exception

### ISSUE THREAD

**Issue by cybex-dev** [flutter/flutter#105019]
**TL;DR**

`Container` (with `padding`|`margin`) -> `Expanded` -> `DefaultTabController` throws exception. Removing `margin` | `padding` resolves this problem.

```dart
Column(
  children: [
    // ...

    Container(
      // uncomment either, both will cause exception
      // padding: const EdgeInsets.all(8.0),
      // margin: const EdgeInsets.all(8.0),
      child: Expanded(
        child: DefaultTabController(
          length: 3,
          // ...
        ),
      ),
    )
  ],
)
```

**Exception:**
```
The following assertion was thrown while applying parent data.:
Incorrect use of ParentDataWidget.

The ParentDataWidget Expanded(flex: 1) wants to apply ParentData of type FlexParentData to a RenderObject, which has been set up to accept ParentData of incompatible type BoxParentData.

Usually, this means that the Expanded widget has the wrong ancestor RenderObjectWidget. Typically, Expanded widgets are placed directly inside Flex widgets.
The offending Expanded is currently placed inside a Padding widget.
```
---

**Expected results:** 
An app, with an app bar, 3 tabs and each tab's content.

**Actual results:**
An app, with an app bar, no tab related content and exception in debug terminal.

**Also Tried:**
- Used `padding` / `margin` in `Container`
- Replaced `Container` with `Padding`
- Removed contained `Expanded`.

**Sample Use-case:**
`TabBar` filling up (`Expanded`) screen below widgets, `Text`, buttons, etc.

![UseCase Image](https://imgur.com/4qdeUF5)

**Image:** https://imgur.com/4qdeUF5

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Demo',
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    const primaryColor = Colors.red;
    return Scaffold(
      appBar: AppBar(title: const Text("Page Title")),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        mainAxisAlignment: MainAxisAlignment.start,
        children: [
          const SizedBox(height: 16),
          Container(
            //  uncomment either to throw exception
            // padding: const EdgeInsets.all(8.0),
            // margin: const EdgeInsets.all(8.0),
            child: Expanded(
              child: DefaultTabController(
                length: 3,
                child: Container(
                  margin: const EdgeInsets.symmetric(horizontal: 10.0),
                  child: Column(
                    children: const <Widget>[
                      TabBar(
                        labelColor: Colors.blue,
                        tabs: <Widget>[
                          Tab(text: 'Tab 1',),
                          Tab(text: 'Tab 2'),
                          Tab(text: 'Tab 3'),
                        ],
                      ),
                      Expanded(
                        child: TabBarView(
                          children: <Widget>[
                            Center(child: Text('Tab View 1')),
                            Center(child: Text('Tab View 2')),
                            Center(child: Text('Tab View 3')),
                          ],
                        ),
                      )
                    ],
                  ),
                ),
              ),
            ),
          )
        ],
      ),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
Performing hot restart...
Waiting for connection from debug service on Chrome...
Restarted application in 167ms.

======== Exception caught by widgets library =======================================================
The following assertion was thrown while applying parent data.:
Incorrect use of ParentDataWidget.

The ParentDataWidget Expanded(flex: 1) wants to apply ParentData of type FlexParentData to a RenderObject, which has been set up to accept ParentData of incompatible type BoxParentData.

Usually, this means that the Expanded widget has the wrong ancestor RenderObjectWidget. Typically, Expanded widgets are placed directly inside Flex widgets.
The offending Expanded is currently placed inside a Padding widget.

The ownership chain for the RenderObject that received the incompatible parent data was:
  Padding ← Container ← _TabControllerScope ← DefaultTabController ← Expanded ← Padding ← Container ← Column ← _BodyBuilder ← MediaQuery ← ⋯
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
packages/flutter/src/widgets/framework.dart 5922:11                           <fn>
packages/flutter/src/widgets/framework.dart 5938:14                           [_updateParentData]
packages/flutter/src/widgets/framework.dart 5961:50                           attachRenderObject
packages/flutter/src/widgets/framework.dart 5635:5                            mount
packages/flutter/src/widgets/framework.dart 6283:11                           mount
packages/flutter/src/widgets/framework.dart 3790:13                           inflateWidget
packages/flutter/src/widgets/framework.dart 3540:18                           updateChild
packages/flutter/src/widgets/framework.dart 4780:16                           performRebuild
packages/flutter/src/widgets/framework.dart 4477:5                            rebuild
packages/flutter/src/widgets/framework.dart 4735:5                            [_firstBuild]
packages/flutter/src/widgets/framework.dart 4729:5                            mount
...


======== Exception caught by rendering library =====================================================
The following assertion was thrown during performLayout():
RenderFlex children have non-zero flex but incoming height constraints are unbounded.

When a column is in a parent that does not provide a finite height constraint, for example if it is in a vertical scrollable, it will try to shrink-wrap its children along the vertical axis. Setting a flex on a child (e.g. using Expanded) indicates that the child is to expand to fill the remaining space in the vertical direction.
These two directives are mutually exclusive. If a parent is to shrink-wrap its child, the child cannot simultaneously expand to fit its parent.

Consider setting mainAxisSize to MainAxisSize.min and using FlexFit.loose fits for the flexible children (using Flexible rather than Expanded). This will allow the flexible children to size themselves to less than the infinite remaining space they would otherwise be forced to take, and then will cause the RenderFlex to shrink-wrap the children rather than expanding to fit the maximum constraints provided by the parent.

If this message did not help you determine the problem, consider using debugDumpRenderTree():
  https://flutter.dev/debugging/#rendering-layer
  http://api.flutter.dev/flutter/rendering/debugDumpRenderTree.html
The affected RenderFlex is: RenderFlex#7a5e8 relayoutBoundary=up4 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
  parentData: offset=Offset(0.0, 0.0) (can use size)
  constraints: BoxConstraints(0.0<=w<=1088.0, 0.0<=h<=Infinity)
  size: MISSING
  direction: vertical
  mainAxisAlignment: start
  mainAxisSize: max
  crossAxisAlignment: center
  verticalDirection: down
...  child 1: RenderCustomPaint#b9d56 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...    parentData: offset=Offset(0.0, 0.0); flex=null; fit=null
...    constraints: MISSING
...    size: MISSING
...    painter: _IndicatorPainter#d52b6(AnimationController#6931e(▶ 0.000; paused))
...    child: _TabLabelBarRenderer#abed2 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...      parentData: <none>
...      constraints: MISSING
...      size: MISSING
...      direction: horizontal
...      mainAxisAlignment: start
...      mainAxisSize: max
...      crossAxisAlignment: center
...      textDirection: ltr
...      verticalDirection: down
...      child 1: RenderSemanticsAnnotations#c0a45 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...        parentData: offset=Offset(0.0, 0.0); flex=1; fit=FlexFit.tight
...        constraints: MISSING
...        size: MISSING
...        child: RenderMouseRegion#01e88 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...          parentData: <none>
...          constraints: MISSING
...          size: MISSING
...          listeners: enter, exit
...          cursor: SystemMouseCursor(click)
...      child 2: RenderSemanticsAnnotations#e7264 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...        parentData: offset=Offset(0.0, 0.0); flex=1; fit=FlexFit.tight
...        constraints: MISSING
...        size: MISSING
...        child: RenderMouseRegion#d7ae3 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...          parentData: <none>
...          constraints: MISSING
...          size: MISSING
...          listeners: enter, exit
...          cursor: SystemMouseCursor(click)
...      child 3: RenderSemanticsAnnotations#bc056 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...        parentData: offset=Offset(0.0, 0.0); flex=1; fit=FlexFit.tight
...        constraints: MISSING
...        size: MISSING
...        child: RenderMouseRegion#14c7a NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...          parentData: <none>
...          constraints: MISSING
...          size: MISSING
...          listeners: enter, exit
...          cursor: SystemMouseCursor(click)
...  child 2: _RenderScrollSemantics#7e2b0 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...    parentData: offset=Offset(0.0, 0.0); flex=1; fit=FlexFit.tight
...    constraints: MISSING
...    semantic boundary
...    size: MISSING
...    child: RenderPointerListener#39906 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...      parentData: <none>
...      constraints: MISSING
...      size: MISSING
...      behavior: deferToChild
...      listeners: signal
...      child: RenderSemanticsGestureHandler#2c038 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...        parentData: <none>
...        constraints: MISSING
...        size: MISSING
...        behavior: opaque
...        gestures: <none>
...        child: RenderPointerListener#aa8f7 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...          parentData: <none>
...          constraints: MISSING
...          size: MISSING
...          behavior: opaque
...          listeners: down
The creator information is set to: Column ← Container ← _TabControllerScope ← DefaultTabController ← Expanded ← Padding ← Column ← Padding ← Container ← _BodyBuilder ← MediaQuery ← LayoutId-[<_ScaffoldSlot.body>] ← ⋯
The nearest ancestor providing an unbounded width constraint is: RenderFlex#e0913 relayoutBoundary=up2 NEEDS-LAYOUT NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
...  parentData: offset=Offset(0.0, 0.0) (can use size)
...  constraints: BoxConstraints(0.0<=w<=1104.0, 0.0<=h<=454.0)
...  size: MISSING
...  direction: vertical
...  mainAxisAlignment: start
...  mainAxisSize: max
...  crossAxisAlignment: center
...  verticalDirection: down

See also: https://flutter.dev/layout/

If none of the above helps enough to fix this problem, please don't hesitate to file a bug:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md
The relevant error-causing widget was: 
  Column Column:file:///Users/mac/AndroidStudioProjects/test_project/lib/main.dart:49:28
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
packages/flutter/src/rendering/flex.dart 926:9                                <fn>
packages/flutter/src/rendering/flex.dart 928:14                               performLayout
packages/flutter/src/rendering/object.dart 1887:7                             layout
packages/flutter/src/rendering/shifted_box.dart 233:5                         performLayout
packages/flutter/src/rendering/object.dart 1887:7                             layout
packages/flutter/src/rendering/layout_helper.dart 56:10                       layoutChild
...

======== Exception caught by rendering library =====================================================
The following assertion was thrown during performLayout():
Assertion failed: file:///Applications/flutter/packages/flutter/lib/src/rendering/box.dart:1982:12
hasSize
"RenderBox was not laid out: RenderFlex#7a5e8 relayoutBoundary=up4 NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE"

The relevant error-causing widget was: 
  Padding Padding:file:///Users/mac/AndroidStudioProjects/test_project/lib/main.dart:42:13
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3    assertFailed
packages/flutter/src/rendering/box.dart 1982:12                               get size
packages/flutter/src/rendering/shifted_box.dart 237:32                        performLayout
packages/flutter/src/rendering/object.dart 1887:7                             layout
packages/flutter/src/rendering/layout_helper.dart 56:10                       layoutChild
packages/flutter/src/rendering/flex.dart 829:43                               [_computeSizes]
packages/flutter/src/rendering/flex.dart 931:32                               performLayout
packages/flutter/src/rendering/object.dart 1887:7                             layout
packages/flutter/src/rendering/shifted_box.dart 233:5                         performLayout
...


======== Exception caught by rendering library =====================================================
The following assertion was thrown during performLayout():
Assertion failed: file:///Applications/flutter/packages/flutter/lib/src/rendering/box.dart:1982:12
hasSize
"RenderBox was not laid out: RenderPadding#1e6c8 relayoutBoundary=up3 NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE"

The relevant error-causing widget was: 
  Column Column:file:///Users/mac/AndroidStudioProjects/test_project/lib/main.dart:30:16
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3    assertFailed
packages/flutter/src/rendering/box.dart 1982:12                               get size
packages/flutter/src/rendering/layout_helper.dart 57:17                       layoutChild
packages/flutter/src/rendering/flex.dart 829:43                               [_computeSizes]
packages/flutter/src/rendering/flex.dart 931:32                               performLayout
packages/flutter/src/rendering/object.dart 1887:7                             layout
...

======== Exception caught by rendering library =====================================================
The following assertion was thrown during performLayout():
Assertion failed: file:///Applications/flutter/packages/flutter/lib/src/rendering/box.dart:1982:12
hasSize
"RenderBox was not laid out: RenderFlex#e0913 relayoutBoundary=up2 NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE"

The relevant error-causing widget was: 
  Container Container:file:///Users/mac/AndroidStudioProjects/test_project/lib/main.dart:28:13
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3    assertFailed
packages/flutter/src/rendering/box.dart 1982:12                               get size
packages/flutter/src/rendering/shifted_box.dart 237:32                        performLayout
packages/flutter/src/rendering/object.dart 1887:7                             layout
packages/flutter/src/rendering/custom_layout.dart 171:10                      layoutChild
packages/flutter/src/material/scaffold.dart 1005:7                            performLayout
...

======== Exception caught by rendering library =====================================================
The following assertion was thrown during performLayout():
Assertion failed: file:///Applications/flutter/packages/flutter/lib/src/rendering/box.dart:1982:12
hasSize
"RenderBox was not laid out: RenderPadding#0c5bf relayoutBoundary=up1 NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE"

The relevant error-causing widget was: 
  Scaffold Scaffold:file:///Users/mac/AndroidStudioProjects/test_project/lib/main.dart:26:12
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3    assertFailed
packages/flutter/src/rendering/box.dart 1982:12                               get size
packages/flutter/src/rendering/custom_layout.dart 172:17                      layoutChild
packages/flutter/src/material/scaffold.dart 1005:7                            performLayout
...
```
</details>

<details>
  <summary>Flutter Analyze</summary>

```
flutter analyze
Analyzing test_project...                                               
No issues found! (ran in 2.7s)
```
</details>

<details>
  <summary>Flutter Doctor</summary>

```
flutter doctor -v
[✓] Flutter (Channel stable, 2.10.5, on macOS 12.3 21E230 darwin-x64, locale en-ZA)
    • Flutter version 2.10.5 at /Applications/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5464c5bac7 (6 weeks ago), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mac/Library/Android/sdk
    • Platform android-32, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] Android Studio (version 2020.3)
    • Android Studio at /Users/mac/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7935034/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3)
    • IntelliJ at /Users/mac/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3)
    • IntelliJ at /Users/mac/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/213.5744.223/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2020.3.2)
    • IntelliJ at /Users/mac/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.64.2)
    • VS Code at /Users/mac/Downloads/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (4 available)
    • sdk gphone x86 (mobile)    • emulator-5554                        • android-x86    • Android 11 (API 30) (emulator)
    • iPhone 12 Pro Max (mobile) • 788ECA65-C187-40C1-A8EB-14E023EA1A19 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-4 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • macOS 12.3 21E230 darwin-x64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 102.0.5005.61

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by maheshmnj** [comment]
Hi @cybex-dev, Thanks for filing the issue. I believe this is working as intended, Expanded widgets are supposed to be immediate child of Row or column  as mentioned in the docs

> An [Expanded](https://api.flutter.dev/flutter/widgets/Expanded-class.html) widget must be a descendant of a [Row](https://api.flutter.dev/flutter/widgets/Row-class.html), [Column](https://api.flutter.dev/flutter/widgets/Column-class.html), or [Flex](https://api.flutter.dev/flutter/widgets/Flex-class.html),

Swapping Expanded with its parent (Container) should resolve the issue.

```dart
 ...
  Expanded(
     child: Container(
        padding: const EdgeInsets.all(8.0),
        margin: EdgeInsets.all(8.0),
        child: DefaultTabController(
 ...

```

<details>
<summary>modified code samplev</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Demo',
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    const primaryColor = Colors.red;
    return Scaffold(
      appBar: AppBar(title: const Text("Page Title")),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        mainAxisAlignment: MainAxisAlignment.start,
        children: [
          const SizedBox(height: 16),
          Expanded(
            child: Container(
              padding: const EdgeInsets.all(8.0),
              child: DefaultTabController(
                length: 3,
                child: Container(
                  margin: const EdgeInsets.symmetric(horizontal: 10.0),
                  child: Column(
                    children: const <Widget>[
                      TabBar(
                        labelColor: Colors.blue,
                        tabs: <Widget>[
                          Tab(
                            text: 'Tab 1',
                          ),
                          Tab(text: 'Tab 2'),
                          Tab(text: 'Tab 3'),
                        ],
                      ),
                      Expanded(
                        child: TabBarView(
                          children: <Widget>[
                            Center(child: Text('Tab View 1')),
                            Center(child: Text('Tab View 2')),
                            Center(child: Text('Tab View 3')),
                          ],
                        ),
                      )
                    ],
                  ),
                ),
              ),
            ),
          )
        ],
      ),
    );
  }
}

```
</details>
